자동차 경주 5단계

bsgreentea opened the pull request:
### 구현
MVC 패턴에 맞게 코드를 재구성했습니다.
테스트 코드를 추가했습니다.

### 질문
RacingGameTest의 '우승자 찾기' 테스트가  RacingGame의 getWinners() 메서드와 큰 차이가 없는 것 같습니다.
이렇게 되면 테스트라기보다는 코드 복붙이랑 다를바가 없을 것 같은데, 어떻게 해야 유의미한 테스트를 작성할 수 있을까요?

bsgreentea committed:
refactor: MVC 패턴에 맞게 패키지를 나눈다.

bsgreentea committed:
refactor: 각 패키지의 역할에 맞게 코드를 수정한다.

- 입출력은 RaceController에서 다루도록 수정
- 자동차 리스트와 게임 횟수는 RacingGame의 생성자에서 받도록 변경

bsgreentea committed:
refactor: 메서드명을 좀 더 알기 쉽게 바꾼다.

- 한 번의 턴을 진행하는 것을 알 수 있도록 drive -> driveOneTurn으로 메서드명 변경

bsgreentea committed:
test: Car 클래스에 대한 테스트 코드를 작성한다.

- 테스트용으로 최대한 진행할 수 있는 게임 회차를 두고, 게임 진행 시 최대 주행거리가 이를 넘어가는지 않는 값이 맞는지 확인

bsgreentea committed:
test: 레이싱 게임에 대한 테스트 코드를 작성한다.

- 한 번의 턴 진행과 우승자 확인에 대한 테스트 코드 작성
- Car의 모든 public 메서드에 대해 테스트하도록 테스트 추가

Rok93 commented on src/main/kotlin/week1/racingcar/controller/RaceController.kt:
```diff
@@ -0,0 +1,24 @@
+        val raceInfo = InputView.getValidateRaceInfo()
+        val racingGame = RacingGame(raceInfo.cars.map { Car(it) }, raceInfo.timesToRace)
+
+        repeat(racingGame.timesToRace) {
```
racingGame 객체에게 게임이 끝났는지? 여부를 물어보면 어떨까요? 🤔

```kotlin
     while(racingGame.isPlatying()) {
            val carsAfterDriving = racingGame.driveOneTurn()
            ResultView.showResult(carsAfterDriving)
     }
```

`driveOneTurn` 메서드를 호출할 때, racingGame 객체의 프로퍼티인 **timesToRace** 값도 변경되어야할 것 같네요. 👀

bsgreentea replied:
아하 timesToRace의 접근 제한자를 private으로 만들 수 있겠군요

Rok93 commented on src/main/kotlin/week1/racingcar/controller/RaceController.kt:
```diff
@@ -0,0 +1,24 @@
+
+    fun startRacing() {
+        val raceInfo = InputView.getValidateRaceInfo()
+        val racingGame = RacingGame(raceInfo.cars.map { Car(it) }, raceInfo.timesToRace)
```
List<Car>를 만드는 로직을 private 메서드로  뽑으면 조금 더 깔끔할 것 같아요. 👀
```suggestion
        val racingGame = RacingGame(createCars(raceInfo.cars), raceInfo.timesToRace)
```

Rok93 commented on src/main/kotlin/week1/racingcar/controller/RaceController.kt:
```diff
@@ -0,0 +1,24 @@
+        val racingGame = RacingGame(raceInfo.cars.map { Car(it) }, raceInfo.timesToRace)
+
+        repeat(racingGame.timesToRace) {
+            val carsAfterDriving = racingGame.driveOneTurn()
+            ResultView.showResult(carsAfterDriving)
```
사소한 부분이지만 `ResultView.showResult` 라고 했으니 메서드 파라미터로 주입되는 변수의 네이밍도 xxxResult 같은 느낌이 됐으면 자연스러웠을 것 같아요.🤔    `driveResult` 같은 네이밍도 좋았을 것 같아요

Rok93 commented on src/main/kotlin/week1/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,20 @@
+        return cars
+    }
+
+    fun getWinners(): List<String> {
```
getXXX와 같은 네이밍을 가급적 지양하는 것이 좋다고 생각합니다. 
RacingGame의 객체가 winners라는 프로퍼티를 가지고있고 이 상태값을 단순히 외부로 가져오는 작업으로 오해할 수 있기 때문이에요. 

실제 이 메서드의 행위를 잘 표현할 수 있도록 네이밍을 수정해보시면 좋을 것 같습니다. 
ex) `우승자를 선별한다`, `우승자를 뽑는다`

Rok93 commented on src/test/kotlin/week1/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,43 @@
+        repeat(DEFAULT_TEST_TIMES) {
+            car1.moveOrNot()
+        }
+        assert(car1.mileage <= DEFAULT_TEST_TIMES)
```
assertJ 라이브러리를 사용하사ㅣ면 조금 더 직관적으로 검증부를 이해할 수 있어요. 

`assertThat(car1.mileage).isLessThanOrEqualTo(MAX_TEST_TIMES)`

또한 테스트가 실패했을 때, 테스트가 왜 실패했는지 조금 더 상세한 내용을 알려줘서 깨진 테스트를 수정하는데 많은 도움을 받을 수 있어요. 😃

<img width="1254" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/6eed5122-e3e7-4040-802f-46b33639d5fb">

<img width="1175" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/85fb4aa3-a76b-4e69-b48e-704540b50083">

bsgreentea replied:
오 의미도 파악하기 쉽고 에러도 더 잘 보여주는군요 👍

Rok93 commented on src/test/kotlin/week1/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,43 @@
+    fun `한 턴 진행`() {
+        val car1 = Car("K5")
+        repeat(DEFAULT_TEST_TIMES) {
+            car1.moveOrNot()
```
이 테스트가 moveOrNot 메서드가 잘 동작하는지를 보장해주고 있다고 생각하시나요? 👀 
(RacingGame의 driveOneTurn 메서드도 동일한 상황인 것 같아요. 👀)

[자동차 경주게임 5단계 미션 소개 페이지 - 테스트 관련 내용](https://edu.nextstep.camp/s/lepg4Qrl/ls/JD7OaEeI) 을 다시 한번 가볍게 훑어보시면 좋을 것 같습니다. 😃

이 테스트가 왜 하기 힘든지 고민을 해보시면 좋을 것 같아요. 
moveOrNot 메서드가 테스트하기 힘든 이유는 테스트하기 힘든 요소가 로직에 포함되어있기 때문입니다. 👀
이 테스트하기 힘든 요소를 외부로 분리하게되면(ex. 메서드 파라미터 or 생성자 파라미터로 주입) 테스트 할 수 있는 영역을 확보할 수 있고 테스트하기 힘들었던 부분 중에서 테스트할 수 있는 영역을 조금 씩 더 늘려갈 수 있어요. 🤗

또한 시간 여유가 되신다면 [해당 영상](https://www.youtube.com/watch?v=YdtknE_yPk4)을 보시는게 도움이 되실 것 같습니다. 🙏

Rok93 commented on src/test/kotlin/week1/racingcar/domain/RacingGameTest.kt:
```diff
@@ -0,0 +1,35 @@
+    @Test
+    fun `우승자 찾기`() {
+        val racingGame = RacingGame(cars)
+        repeat(10) {
+            val carsAfterDriving = racingGame.driveOneTurn()
+            carsAfterDriving.maxOfOrNull { it.mileage }?.let { farthestMileage ->
+                val leadingCars = carsAfterDriving.filter { it.isSameMileage(farthestMileage) }
+                assert(leadingCars.map { it.name }.toSet() == racingGame.getWinners().toSet())
+            }
```
> RacingGameTest의 '우승자 찾기' 테스트가 RacingGame의 getWinners() 메서드와 큰 차이가 없는 것 같습니다.
이렇게 되면 테스트라기보다는 코드 복붙이랑 다를바가 없을 것 같은데, 어떻게 해야 유의미한 테스트를 작성할 수 있을까요?

이 질문을 해주셨는데, 

저라면 Car 객체의 주 생성자를 아래와 같이 `이름`, `위치`를 다 주입받을 수 있도록 했을 것 같고 

```kotlin
class Car(
    val name: String,
    var mileage: Int = DEFAULT_MILEAGE,
) {
```

테스트 코드는 아래와 같이 변경했을 것 같아요. 
```kotlin
        // given
        val cars = listOf(Car("1", 1), Car("2", 2), Car("3-1", 3), Car("3-2", 3))
        val sut = RacingGame(cars)

        // when
        val actual = sut.getWinners()

        // then
        assertThat(actual).hasSize(2)
        assertThat(actual).containsExactlyElementsOf(listOf("3-1", "3-2"))
```

bsgreentea replied:
테스트코드가 훨씬 깔끔해지는군요. 그런데 mileage를 Car 외부에서도 변경할 수 있게 되는게 좀 고민됩니다.

bsgreentea replied:
음 private하게 만들고 현재 mileage를 반환하는 메서드를 하나 만들어주면 되겠군요!

Rok93 reviewed (changes requested):
보성님 5단계도 빠르게 잘 구현해주셨습니다 👍💯
MVC 패턴으로 잘 분리해주셔서 이 부분은 잘 구현해주신 것 같아 코멘트드릴 부분이 없었어요. 🤩😉

대신 테스트 관련된 코멘트를 추가적으로 달아두었고 이 부분 확인해서 변경해주시면 될 것 같습니다.
😃 특히 "테스트하기 어려운 코드를 어떻게 테스트할 것인가?"를 고민해보시면 좋을 것 같아요. 🙏

미션 진행하면서 어려운 부분이있으면 언제든지 DM 주세요! 😉

bsgreentea committed:
refactor: 5단계 리뷰를 반영한다.

- RacingGame의  timesToRace의 접근 제한자를 private으로 변경
- 경기 중인지를 판단하는 것은 timesToRace에 직접 접근하는 것 대신 isPlaying() 메서드로 대체
- 주행거리에 대한 값은 currentMileage()로 대체
- 우승자 반환하는 메서드의 이름을 getter와 헷갈리지 않도록 filterWinners로 변경
- mileage를 Car의 생성자로 받을 수 있도록 수정
- 테스트 로직이 명확하도록 우승자 찾기 테스트 변경
- 랜덤 생성한 값마다 주행 여부를 모두 판단하도록 테스트 수정
- 랜덤값에 대한 부분을 테스트 가능하게 하기 위해 파라미터로 주입받음

Rok93 commented on src/main/kotlin/week1/racingcar/domain/Car.kt:
```diff
@@ -1,17 +1,16 @@
 ) {
-    var mileage = DEFAULT_MILEAGE
-        private set
 
     init {
         validateCarInfo()
     }
 
+    fun currentMileage() = mileage
```
메서드를 만드시는 것도 괜찮지만 아래와 같이 읽기전용 프로퍼티를 만드시는 것도 괜찮은 방식 같아요. 😃

```suggestion
    val currentMileage
        get() = mileage
    
    init {
        validateCarInfo()
    }
```

이렇게 작성했을 때, java로 변경해보면 어떻게되는지 decompile 해보시면 좋을 것 같아요. 😉
<img width="1510" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/36f8397e-263b-4492-9618-2b34d6d32f1f">

bsgreentea replied:
역할이 딱히 없어서 좀 망설였던 부분인데, 읽기 전용 프로퍼티로 만드는게 보기가 더 좋은 것 같네요!
동작은 완벽히 같군요

Rok93 commented on src/main/kotlin/week1/racingcar/domain/Car.kt:
```diff
@@ -21,17 +20,17 @@ class Car(
 
-    fun moveOrNot() {
-        if (isPossibleToDrive()) mileage++
+    fun moveOrNotByNum(randomNum: Int) {
```
자동차 객체의 입장에서 파라미터로 주입받는 값이 `random` 인 것을 알 필요가 있을까요? 🤔
현실 세계에서 차량이 전진 혹은 정지하는데에 어떤 요소가 필요한지 생각해보고 적절한 네이밍을 지어볼 수 있을 것 같아요. 

ex) condition, fuel(= 연료), gauge(이건 적절한지 모르겠네요 🤔)

Rok93 commented on src/main/kotlin/week1/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,27 @@
+    private val cars: List<Car>,
+    private val timesToRace: Int = 0,
+) {
+    private var racedTime = 0
```
또 하나의 프로퍼티를 선언하셨군요? 
지금의 경우에는 readTime을 하나씩 증가시키는 대신에 timesToRace를 1씩 감소시키는 방법으로도 충분히 대체할 수 있을 것 같아요. 😃  가급적 객체의 프로퍼티는 정말 필요한 경우에만 추가해주세요. 😉

Rok93 commented on src/main/kotlin/week1/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,27 @@
+) {
+    private var racedTime = 0
+
+    fun driveOneTurn(): List<Car> {
```
아마 `driveOneTunr` 메서드도 이전의 `moveOrNot` 메서드처럼 테스트하기 힘들 것 같아요. 😃
moveOrNot을 테스트하기 쉬운 메서드로 만들었던 것 처럼 driveOneTurn도 테스트할 수 있도록 변경해보고 테스트도 조금 더 유의미한 테스트를 작성하도록 해볼까요? 🤗

Rok93 commented on src/main/kotlin/week1/racingcar/domain/RacingGame.kt:
```diff
@@ -0,0 +1,27 @@
+        return cars
+    }
+
+    fun isPlaying() = racedTime < timesToRace
```
isPlaying 메서드도 마찬가지로 테스트 코드를 작성해주시면 좋을 것 같습니다. 🙏

Rok93 commented on src/test/kotlin/week1/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,36 @@
+import week1.racingcar.domain.Car.Companion.RANDOM_NUM_RANGE
+
+class CarTest {
+    @Test
+    fun `생성자 확인`() {
+        val car1 = Car("James")
+        car1.name shouldBe "James"
+
+        assertThrows(IllegalArgumentException::class.java) {
+            Car("Thomas")
+        }
+    }
```
하나의 테스트 케이스에 대해서만 테스트를 작성해주세요., 

지금의 경우에는 1. 정상적으로 Car 객체를 생성하는 케이스  2. 글자 수 초과로 객체 생성 실패 케이스와 같이 총 2개의 테스트가 하나의 테스트 케이스에 존재하고 있어요. 

또한 공백에 의해서 Car 객체 생성을 실패하는 케이스에 대해서도 테스트를 작성하시면 좋을 것 같습니다. 😉

Rok93 commented on src/test/kotlin/week1/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,36 @@
+
+    @Test
+    fun `주행거리 비교`() {
+        val car = Car("EV6")
+        car.isSameMileage(0)
```
사소하지만 주생성자로 `mileage`를 지정할 수 있으니 0 말고 다른 숫자로 검증해보면 어떨까요?

Rok93 commented on src/test/kotlin/week1/racingcar/domain/CarTest.kt:
```diff
@@ -0,0 +1,36 @@
+        (0 until RANDOM_NUM_RANGE).forEach {
+            val car = Car("K5")
+            car.moveOrNotByNum(it)
+            val expectedMileage = if (it < MIN_NUM_TO_GO) 0 else 1
```
지금 같은 테스트 케이스는 테스트 안에 (마치 비즈니스 로직처럼) 분기문 처리를 하는 것 보다는  여러가지 인자에 대해서 테스트를 하는 것이 더 좋을 것 같아요. 😃 
[Junit의 Paramerized Test](https://www.baeldung.com/parameterized-tests-junit-5)를 활용해서 0 ~ 9 까지의 숫자에 대해서 어떤 예상한 값이 나오는지에 대해 테스트를 작성해보셨으면 합니다. 🙏

추가적으로 아래의 글을 읽어보시면 좋을 것 같습니다. 😉
[조건부 (if~else) 로직 테스트코드에서 사용하지 않기](https://jojoldu.tistory.com/642)

Rok93 commented on src/test/kotlin/week1/racingcar/domain/RacingGameTest.kt:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    companion object {
+        private val carsBeforeDriving = listOf(
+            Car("1"),
+            Car("2"),
+            Car("3"),
+            Car("4"),
+        )
+
+        private val carsAfterDriving = listOf(
+            Car("1", 0),
+            Car("2", 10),
+            Car("3", 3),
+            Car("4", 7),
+            Car("5", 10),
+        )
```
두 가지 피드백을 드리고싶습니다. 👀

1. 먼저 상수 선언은 Kotlin Code Convention에 따라서 UPPER_CAMEL_CASE로 명명하라고 가이드하고있으니 변경해주시면 좋을 것 같습니다. 

2. 한번 지정하고 여러 테스트 케이스에서 재활용해서 사용할 수 있는 경우라면 지금처럼 상수나 프로퍼티로 선언해두고 재사용하는 것도 괜찮지만 지금의 경우에는 각 테스트 케이스마다 다른 값들을 사용하고 있는 것으로 보입니다.  이런 경우에는 그냥 테스트 케이스 내부에서 로컬 변수로 선언하고 테스트를 진행하시는 것이 좋을 것 같습니다.

Rok93 reviewed (changes requested):
보성님 피드백 반영 잘해주셨어요 👍👍 
테스트하기 힘들었던 moveOrNot 메서드를 테스트하기 쉽게 만들었던 것 처럼 
driveOneTurn 메서드도 테스트하기 쉬운 테스트로 변경해서 테스트를 해보시면 좋을 것 같습니다. 
그 외에도 테스트 관련 코멘트들 남겨두었으니 확인해서 반영해주시면 좋을 것 같습니다. 🙏
이제 진짜 자동차 미션 마지막이 보이는 것 같으니 조금만 더 힘내세요 💪💪

bsgreentea committed:
refactor: 5단계 리뷰를 반영한다.

- 현재 주행거리를 가져오는 용도의 currentMileage 메서드 대신 읽기 전용 프로퍼티를 사용하도록 수정
- moveOrNotByNum 메서드의 파라미터명이 클래스 의미에 조금 더 적합하게끔 수정
- racedTime의 역할을 timesToRace로 대체함(불필요 변수 줄이기)
- 반복 사용하지 않는 리스트를 메서드 내에서 선언하기
- 게임 진행이 가능한지 판단하는 isPlaying 메서드에 대한 테스트 코드 작성
- 자동차 객체 생성의 정상적 케이스와 비정상적 케이스에 대한 테스트코드를 분리
- 랜덤 넘버에 따른 주행 여부 테스트 코드에 MethodSource를 사용하고, 해당 메서드에 넘벼별 상황이 나타나도록 함
- driveOneTurn에 대한 테스트코드 변경(currentMileage에 대한 테스트는 CarTest에서 진행중이므로 게임 횟수가 줄어들었는지 확인하는게 적절할 것 같다.)

bsgreentea commented:
https://github.com/next-step/kotlin-racingcar/pull/1518#discussion_r1384833772

수정하다보니 게임 횟수가 줄어들었는지 확인하는 것 외에는 테스트할 부분이 딱히 없는 것 같습니다.
해당 메서드 테스트에 있어 더 적절한 방법이 있었을까요..??

Rok93 commented on src/test/kotlin/week1/racingcar/domain/RacingGameTest.kt:
```diff
@@ -0,0 +1,51 @@
+        )
+        val timesToRace = 2
+        val racingGame = RacingGame(carsBeforeDriving, timesToRace)
+        racingGame.driveOneTurn()
+        racingGame.isPlaying() shouldBe true
+        racingGame.driveOneTurn()
+        racingGame.isPlaying() shouldBe false
```
> 수정하다보니 게임 횟수가 줄어들었는지 확인하는 것 외에는 테스트할 부분이 딱히 없는 것 같습니다.
해당 메서드 테스트에 있어 더 적절한 방법이 있었을까요..??

위와 같은 질문을 주셨는데, 저는 `driveOneTurn`메서드를 호출함으로써 Cars 값의 위치 값이 변경되었을거고(아닌 경우도 있겠지만요 
 😃) 이 변경된 Cars의 각 위치 값이 의도한 값이 맞는지를 검증했을 것 같아요. 

하지만 지금은 이런 테스트를 할 수 없었을것 같습니다. 그 이유는 driveOneTurn() 메서드 내부에서는 Random한 값에 의존하는 기능이 있기 때문이죠 👀   이 값을 (car 객체의 moveOrNotByNum 처럼) 메서드 파라미터로 뽑아내보면 좋겠다는 의견이었어요. 😃

bsgreentea replied:
아하 그렇군요! driveOneTurn()은 게임 진행 횟수 체크를 제외하면 moveOrNotByNum() 테스트로 어느정도 커버가 되었다고 생각했습니다. 감사합니다!

Rok93 reviewed (approved):
보성님 피드백 반영 잘해주셨습니다. 👍👍👍
이제 자동차 경주 게임 미션은 충분히 잘 수행해주신 것 같아서, 이만 종료하려고합니다. 😉
그동안 정말 고생많으셨고, 남은 미션들도 꼭 완주하셨으면 좋겠습니다. 💪💪
